src/platform/plugins/shared/discover/public/application/main/components/layout/cascaded_documents/blocks/cascade_leaf_component.tsx

```diff
@@ -0,0 +1,337 @@
+      () =>
+        getRenderCustomToolbarWithElements({
+          leftSide: (
+            <EuiText size="s">
+              <b>
+                <FormattedMessage
+                  id="discover.dataCascade.row.cell.toolbar.heading"
+                  defaultMessage="{count, plural, =0 {no results} =1 {1 result} other {# results}}"
+                  values={{ count: cellData.length }}
+                />
+              </b>
+            </EuiText>
+          ),
+        }),
+      [cellData]
+    );
+
+    const setExpandedDocFn = useCallback(
+      (...args: Parameters<NonNullable<UnifiedDataTableProps['setExpandedDoc']>>) =>
+        setExpandedDoc(args[0]),
+      [setExpandedDoc]
+    );
+
+    const renderCustomCascadeGridBodyCallback = useCallback(
+      ({
+        Cell,
+        visibleColumns,
+        visibleRowData,
+        setCustomGridBodyProps,
+        gridWidth,
+        headerRow,
+        footerRow,
+      }: EuiDataGridCustomBodyProps) => (
+        <CustomCascadeGridBodyMemoized
+          key={isCellInFullScreenMode ? `full-screen-${cellId}` : cellId}
+          Cell={Cell}
+          data={cellData}
+          visibleColumns={visibleColumns}
+          visibleRowData={visibleRowData}
+          headerRow={headerRow}
+          footerRow={footerRow}
+          gridWidth={gridWidth}
+          getScrollMargin={getScrollMargin}
+          setCustomGridBodyProps={setCustomGridBodyProps}
+          getScrollElement={getScrollElement}
+          initialOffset={getScrollOffset}
+          isFullScreenMode={isCellInFullScreenMode}
+        />
+      ),
+      [cellData, cellId, getScrollElement, getScrollMargin, getScrollOffset, isCellInFullScreenMode]
+    );
+
+    return (
+      <EuiPanel paddingSize="none">
+        <UnifiedDataTable
+          dataView={dataView}
+          showTimeCol={showTimeCol}
+          showKeyboardShortcuts={showKeyboardShortcuts}
+          services={services}
+          sort={EMPTY_SORT}
+          isSortEnabled={false}
+          enableInTableSearch
+          ariaLabelledBy="data-cascade-leaf-cell"
+          consumer={`discover_esql_cascade_row_leaf_${cellId}`}
+          rows={cellData}
+          loadingState={DataLoadingState.loaded}
+          columns={selectedColumns}
+          onSetColumns={setSelectedColumns}
+          sampleSizeState={sampleSize}
+          renderCustomToolbar={renderCustomToolbarWithElements}
+          onUpdateSampleSize={setSampleSize}
+          expandedDoc={expandedDoc}
+          setExpandedDoc={setExpandedDocFn}
+          dataGridDensityState={cascadeDataGridDensityState}
+          onUpdateDataGridDensity={setCascadeDataGridDensityState}
+          renderDocumentView={renderDocumentView}
+          renderCustomGridBody={renderCustomCascadeGridBodyCallback}
+          onFullScreenChange={setIsCellInFullScreenMode}
+          externalCustomRenderers={externalCustomRenderers}
+          paginationMode="infinite"
```

Comment:
Can we switch to pagination with pages? Otherwise, the scrolling is not smooth and it's making it harder to expand the next category.